Sources/SwiftSyntax/SyntaxNodes.swift.gyb

```diff
@@ -66,8 +61,7 @@ public protocol ${node.name}: Syntax {}
 %     for line in dedented_lines(node.description):
 /// ${line}
 %     end
-/// Syntax nodes are not thread-safe.
-public struct ${node.name}: ${base_type}, _SyntaxBase, Hashable {
```

Comment:
I think we don't need to give up `Equatable`/`Hashable` with this patch. Could we use the absolute position from the beginning of the file as the identity?

```swift
public static func ==(lhs: Syntax, rhs: Syntax) -> Bool {
  return base.position
}

public func hash(into hasher: inout Hasher) {
  hasher.combine(base.position)
}
```

This exposes an invariant that you can't rely on `==`/`hashValue` between trees, which is unfortunate. But it's par for the source with SwiftSyntax -- nodes belonging to different trees have always broken assumptions about the questions you can ask. For example, calling `.sourceLocation` for two nodes from two different trees will give you views into wildly different contexts, and that's no different here.

Reply:
Oooh I have plenty of thoughts about this, I'd love to get your feedback 🙂.

The absolute position can work sufficiently if you check for identity of `TokenSyntax`s since a present TokenSyntax is guaranteed to have a unique position. For example it would work for swift-format's only use-case of [using Syntax as Hashable](https://github.com/google/swift/blob/abb642710435be7f12a0e404dfaef1ac5f3d068c/Sources/SwiftFormatPrettyPrint/TokenStreamCreator.swift#L21).
The other use I saw for Syntax as dictionary key was with SwiftSyntax's own `SyntaxClassifier`, whose use of a Dictionary did not seem important for its functionality and I changed it in https://github.com/apple/swift-syntax/pull/60. (I also intend to change it further to a simpler sequence but this is another PR).

However, position does not provide uniqueness in other nodes in the tree since a parent node has the same position as its first child. You could maybe combine the position with the syntax kind but it's not clear to me whether a `(position, syntax kind)` tuple guarantees uniqueness or not.

An alternative I considered was that we could have an `indexInTree: Int` property that would represent the unique index of the node in its root tree. This could be done efficiently by:
* Having `RawSyntax` store not just its `SourceLength` but also how many total nodes it represents.
* Add `indexInTree` in `AbsoluteSyntaxInfo` and update it appropriately while walking the tree:
```
struct AbsoluteSyntaxInfo {
  let offset: UInt32
  let indexInParent: UInt32
  let indexInTree: UInt32
}
```
(This would make `AbsoluteSyntaxInfo` larger than a `UInt64` but we could also consider not bothering to keep track of `indexInParent`, which is only useful for `replaceSelfWith` modifications.)

Now, to clarify a bit why I tentantively chose to remove the `Equatable/Hashable` conformance:
* It's not clear to me how important it is to provide a unique identity for each node. For `TokenSyntax` the position is sufficient and adapting for swift-format is simple. Perhaps we could make this change and see in practice if there are compelling use cases for SwiftSyntax clients for a unique node identity. If it proves important we can look into adding `indexInParent`.
* Even with having an `indexInParent` property available I have concerns with presenting Syntax nodes as `Equatable/Hashable`. It seems to me it would confuse people over what a Syntax node represents, which is a view of a node in a particular tree configuration and comparing nodes *directly* is not sensible unless they came from the same tree. I feel that using a node's position, or indexInTree, instead makes things much more clear. For example, if I see code doing:
```
if node1.position == node2.position {...}
```
or
```
if node1.indexInTree == node2.indexInTree {...}
```
it is obvious in the code what are the expectations of that comparison, they are properties describing how a node relates in a specific tree. There may even be use cases to compare positions from different trees.

Otherwise with
```
if node1 == node2 {...}
```
The semantics of this kind of comparison are completely non-obvious. What can I infer from this about what it means to compare nodes from different trees ? I think the intuitive assumption with such a code would be that each node is unique and every node from a tree would compare differently from another tree, which is not the case now; it may lead people to make incorrect use of the Syntax nodes.

Reply:
> future extensions of syntax tree such as semantic model, where we associate more information with a syntax node as side table

Let's say we have syntax nodes associated with semantic info; then when you make any change to a syntax tree it is very likely that other syntax nodes would have their semantic information invalidated. I don't think it is correct to assume that because a syntax node had some semantic info associated with in a particular tree, that you can just keep the same semantic info for that syntax node in a different tree configuration.

In other words, I think this particular use case would not benefit from unique identity across trees.

Reply:
I agree for clients like swift-formatter, there're no need to differentiate `this exact syntax node` and `syntax node of this kind in general`. But once we start to provide more context-sensitive info in the swift-syntax APIs, the uniqueness of each syntax node starts to become necessary IMO.

Reply:
> Let's say we have syntax nodes associated with semantic info; then when you make any change to a syntax tree it is very likely that other syntax nodes would have their semantic information invalidated.

That's a good point. But it's equally likely that some nodes' semantic information stays stable so we can re-use it without performing unnecessary work. Whether using cache or re-calculating the semantic information of a node should be implementation detail of the semantic model. But it seems to me we'll always have to recalculate semantic information without being hashable.

Reply:
> Having a unique identifier for each node

To clarify a bit more here, the previous model of Syntax nodes was that every node was unique, and every tree had nodes unique from any other tree. You could not use the node identity to transfer information from one tree to another, all the nodes of the new tree would have completely different identities. Even the raw nodes that got reused during reparsing ended up wrapped by whole new nodes with different identities.

To transfer info from one tree to another you would have to use another property, like perhaps source ranges, unique node identity is not helping you.

Reply:
Here are some more thoughts. If we want unique node identify across trees here's what we can do:
* Add `indexInTree`
* Add an atomically incremented counter for every root node that gets created.
Then we have something like:
```
struct AbsoluteSyntaxInfo {
  let offset: UInt32
  let nodeId: NodeIdentity
}

struct NodeIdentity {
  let rootId: UInt32
  let indexInTree: UInt32
}
```
`rootId` is unique for each root across the process, `indexInTree` is unique within a root tree. With a unique node identity across trees, making nodes `Equatable/Hashable` would be sensible I think.

Reply:
That could definitely work. It's unfortunate that it's no longer a word, but I think having this, and re-enabling `Equatable`/`Hashable` would be great in the long run.

Reply:
> * Add `indexInTree`
> * Add an atomically incremented counter for every root node that gets created.
>   Then we have something like:
This sounds a great approach for me!

Reply:
Please take a look into the latest commit.

Reply:
Thank you, @akyrtzi!